### DISCUSSION TITLE
Configurable numeric equality

### DISCUSSION THREAD

**Discussion by trexfeathers** [SciTools/iris#5987] [Ideas]
_From offline discussion between @trexfeathers, @bjlittle, @pp-mo_

Just as with [metadata equality](https://scitools-iris.readthedocs.io/en/stable/further_topics/lenient_metadata.html#lenient-equality), Iris users have diverse needs when it comes to numeric equality[^1]:

- Precise equality? [Example](https://github.com/SciTools/iris/pull/5926/files#diff-5ef6dd041176be22c0fbce9226150d5f2c0af25e935d34dcf21004c47b7dfb8aR446-R466)
- Allow for floating point differences? [Example](https://github.com/SciTools/iris/blob/6adac1b6f83baa2e1d4a13ced0d6748ff91092ee/lib/iris/cube.py#L3936-L3941)
  - If so, how much relative/absolute tolerance? See [`np.allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html)
- How should `NaN` be handled? [Example](https://github.com/SciTools/iris/blob/6adac1b6f83baa2e1d4a13ced0d6748ff91092ee/lib/iris/util.py#L390-L394)

These considerations touch all corners of Iris: arithmetic, merging (and therefore loading), constraints (I remember #4744), and many more. Not only is this about whether numbers are equal, but also performance (e.g. #5926).

### Problems

- Iris is inconsistent on the type of equality, depending on the operation
- Iris is often opaque about which type of equality is used where
- It is rare for a user to have control over the type of equality used, since it is often used in internal decision making (Can this `Cube` be subtracted from that `Cube`? Are these `Coord`s duplicates? Etcetera)

### Suggestion

Given different user needs, it would probably be unwise to enforce a consistent approach everywhere. User configuration seems the better approach, but since this is often about Iris internals I personally favour a global config with a context manager. We have established a precedent with [Leniency](https://scitools-iris.readthedocs.io/en/stable/further_topics/lenient_maths.html#control-the-behaviour), although this would not look exactly the same since it seems overly complex to offer the config as a parameter for every affected operation.

#### Possible settings

Allow the user to set any of the kwargs in [`np.allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html). Perhaps if nothing is set then **exact** equality is used - possibly using a fast hashing algorithm.

[^1]: I am mostly referring to arrays, but this may apply to single numbers such as those in a [`CoordSystem`](https://github.com/SciTools/iris/blob/6adac1b6f83baa2e1d4a13ced0d6748ff91092ee/lib/iris/coord_systems.py#L48) subclass.

**Comment by trexfeathers** [comment]
Ping @larsbarring

**Reply by larsbarring** [reply]
Thanks for picking this up!  I like the general idea of first using fast hashing (allowing for NaNs) -- after all we expect datasets to be well-behaved and usually they are. But if not, then do something more *lenient* (an possibly more time-consuming). BTW, as you write that this touches all corners of Iris, is it still confined "only" to coordinates and `CoordSystem` data (it is as far as we have seen)?

**Reply by trexfeathers** [reply]
I mean that equality _in general_ touches all corners of Iris, and there is a lot of inconsistency when it comes to NaNs and floating point tolerance.
